servers/rendering/shader_converter.cpp

```diff
@@ -0,0 +1,2591 @@
+				break;
+			} else if (cur_tok == func.body_end_pos) {
+				curr_func = "<global>";
+				break;
+			}
+		}
+		if (cur_tok->get().type == TT::TK_STRUCT) {
+			if (!_skip_struct()) {
+				return false;
+			}
+			continue;
+		}
+
+		switch (cur_tok->get().type) {
+			case TT::TK_FLOAT_CONSTANT: {
+				String const_str = get_token_literal_text(cur_tok->get()).to_lower();
+				// 3.x float constants allowed for a value without a decimal point if it ended in `f` (e.g. `1f`).
+				if (const_str.ends_with("f") && const_str.find(".") == -1 && const_str.find("e") == -1) {
+					return true;
+				}
+			} break;
+			case TT::TK_RENDER_MODE: {
+				while (true) {
+					TokenE *next_tk = get_next_token();
+					if (next_tk->get().type == TT::TK_IDENTIFIER) {
+						String id_text = get_token_literal_text(next_tk->get());
+						if (is_renamed_render_mode(shader_mode, id_text) || has_removed_render_mode(shader_mode, id_text)) {
+							return true;
+						}
+					} else {
+						COND_LINE_MSG_FAIL(next_tk->get().type != TT::TK_COMMA && next_tk->get().type != TT::TK_SEMICOLON, next_tk->get().line, "Invalid render mode declaration");
+					}
+					if (next_tk->get().type == TT::TK_SEMICOLON) {
+						break;
+					}
+				}
+			} break;
+			case TT::TK_IDENTIFIER: {
+				String id = get_token_literal_text(cur_tok->get());
+				if (has_builtin_rename(shader_mode, id, curr_func) || is_removed_builtin(shader_mode, id, curr_func)) {
+					if ((!scope_has_decl(curr_func, id) && !function_decls.has(id))) {
+						return true;
+					}
+				} else if (has_removed_type(id) && peek_next_tk_type() == TT::TK_IDENTIFIER) {
+					// Declaration with unported 3.x type.
+					return true;
+				}
+			} break;
+			default:
+				break;
+		}
+	}
+	return is_3x;
+}
+
+String ShaderDeprecatedConverter::get_error_text() const {
+	return err_str;
+}
+
+bool ShaderDeprecatedConverter::_check_deprecated_type(TokenE *p_type_pos) {
+	if (p_type_pos->get().type == TT::TK_IDENTIFIER && has_removed_type(get_token_literal_text(p_type_pos->get()))) {
+		const String i_err_msg = vformat(RTR("Deprecated type '%s' is not supported by this version of Godot."), get_token_literal_text(p_type_pos->get()));
+		COND_LINE_MSG_FAIL(fail_on_unported, p_type_pos->get().line, i_err_msg);
+		_add_comment_before(i_err_msg, p_type_pos);
+	}
+	return true;
+}
+
+ShaderDeprecatedConverter::TokenE *ShaderDeprecatedConverter::_rename_keyword_id(TokenE *pos, bool p_detected_3x, const HashMap<TokenType, String> &p_new_reserved_word_renames) {
+	String rename = p_new_reserved_word_renames[pos->get().type];
+	reset_to(pos);
+	return replace_curr(mkTok(TT::TK_IDENTIFIER, rename), "Identifier '%s' is a reserved word in this version of Godot, renamed to '%s'.");
+}
+
+bool ShaderDeprecatedConverter::_handle_new_keyword_rename(TokenE *pos, bool p_detected_3x, HashMap<TokenType, String> &p_new_reserved_word_renames) {
+	TokenType tk_type = pos->get().type;
+	String p_name = get_token_literal_text(pos->get());
```

Comment:
```suggestion
	String name = get_token_literal_text(pos->get());
```
Or something similar, `p_` is reserved for arguments